Overwrite metadata in PreUploadCreateCallback if metadata is available in the hook response

jonaskaltenbachz opened the pull request:
This pull request allows metadata returned by the pre-create hook in the "Upload-Metadata" header to be processed in the PreUploadCreateCallback. If the pre-create hook returns metadata in the "Upload-Metadata" header, the PreUploadCreateCallback sets this metadata in the FileInfo object and overwrites the original data. 

This allows metadata from the pre create hook to be modified. 

If no metadata comes back from the hook the original data is used.

See also this question: https://github.com/tus/tusd/issues/811

jonaskaltenbachz commented:
Hello,
I just wanted to ask if there is an update on this topic.

Many greetings,
Jonas

Acconut commented:
Hi Jonas, thank you very much for this PR. As mentioned before, we are interested in adding functionality to modify the upload ID and meta data in the pre-create hook. I am not sure if this will be included in the first proper release of v2 or only later.

While your approach works, I was more thinking about utilizing the new HookResponse struct to add a field for specifying additional meta data that the pre-create hook wants to set: https://github.com/tus/tusd/blob/aff352c413066bc5bfa53f6998024b5ef29240f3/cmd/tusd/cli/hooks/hooks.go#L33

This way, the feature is more explicit, discoverable by users and converted by type checks. What do you think?

jonaskaltenbachz commented:
Thanks for your answer. Adjusting the HookResponse is of course also a good solution. I chose the solution based on the header because I wanted to use as much existing code from you as possible to have as few changes as possible in my pull request. 
But maybe the two solutions are not mutually exclusive. In a first solution, the metadata could be changed based on the response header. And in the future, also via the HookResponse if you adjusted it.

jonaskaltenbachz commented:
Hello @Acconut 

are there any news from your side regarding this topic?

Many greetings,
Jonas

jonaskaltenbachz commented:
Hello @Acconut 
I have changed the pull request as you suggested and added an "AdditionalMetaData" to the "HookResponse". I hope the solution corresponds to your idea? 

Many greetings,
Jonas

jonaskaltenbachz commented:
Hi @Acconut 
I just want to ask if you have already found time to look at the new solution.

jonaskaltenbachz commented:
Hi @Acconut 
I just want to ask if you have already found time to look at the new solution.

jonaskaltenbachz commented:
Hi @Acconut 
I just want to ask if you have already found time to look at the new solution.

jonaskaltenbachz commented:
Hi,
a few months ago I updated the pull request as you suggested and added an "UpdatedMetaData" to the "HookResponse". Could you please add you thoughts regarding this? Or is there any other update from your side? 

It would be really helpful for us if we could merge this pull request.

Thank you in advance and many greetings,
Jonas

Acconut commented on pkg/handler/config.go:
```diff
@@ -54,10 +54,11 @@ type Config struct {
 	// validation of upload metadata etc. Furthermore, HTTPResponse will be ignored and
 	// the error value can contain values for the HTTP response.
-	PreUploadCreateCallback func(hook HookEvent) (HTTPResponse, error)
+	PreUploadCreateCallback func(hook HookEvent) (HTTPResponse, map[string]string, error)
```
I am currently thinking about how we can keep this maintainable in the future. Your PR is interested in modifying the meta data but in the near future we also want to modify the upload ID, storage location etc. I think we should add some struct to hold these potential values here. What do you think?

genofire replied:
an alternative implementation, i wrote here #916 - that should solve ID and Metadata

jonaskaltenbachz replied:
Hi @Acconut 
in https://github.com/tus/tusd/pull/916 you wrote you "would like to avoid merging PR". 
Does this also relate to this pull request? Or should we implement your suggested change and add some struct to hold the return values?

Acconut replied:
Frankly, I am not entirely sure about #916. Through this pointer, it allows the entire struct to be modified, but some of these properties must not be modified by the hook (i.e. Offset, IsFinal, IsPartial). We could just state this rule in the documentation, but it would also be nice if we can enforce it in the API itself.

> in https://github.com/tus/tusd/pull/916 you wrote you "would like to avoid merging PR".

Let me rephrase. I do not want to merge a PR, which addresses some part of the problem of modifying the upload through the pre-create hook, without having the bigger picture on how we can provide this feature with all of its aspects. So if we agree on a way that any kind of hooks can modify relevant properties in a safe manner, I am happy to merge PRs implementing this.

jonaskaltenbachz replied:
Hi @Acconut 
I adjusted the pull request. Please take a look at it and check it out.

genofire commented on pkg/handler/hooks.go:
```diff
@@ -12,6 +12,12 @@ type HookEvent struct {
 
+// HookResponse is the response after a hook is executed.
+type HookResponse struct {
+	// Updated metadata which can be set from the pre create hook and is then used instead of the initial metadata.
```
Please add `ID` here.

jonaskaltenbachz replied:
I added UpdatedID to handler.hookResponse. 

Also fine for you @Acconut ? 
We could move this field also to the proposed handler.UpdatedFileInfo for sure.

jonaskaltenbachz commented:
> Thanks for the updates. I think the structs are getting a bit confusing right now. In v2, we have
> 
> * handler.HTTPRequest
> * handler.HTTPResponse
> * handler.HookEvent (which includes handler.HTTPRequest)
> * hooks.HookRequest (which includes handler.HookEvent)
> * hooks.HookResponse (which includes handler.HTTPResponse)
> 
> Your PR now also introduces a handler.HookResponse, which is not directly connected to any of the previous types. That is a bit too much and even confuses me. I am not saying that we shouldn't have a handler.HookResponse (we already have a handler.HookEvent), but we should think careful on how to lay out the structs here.
> 
> However, I don't have a good proposal right now. What do you think?

What about rename handler.HookResponse to handler.UpdatedFileInfo? Then we don't need the "updated" term in "UpdatedMetaData" and could name it "MetaData".

jonaskaltenbachz commented:
Hi @Acconut 
any update from your side on this topic? Should I implement the renaming to "handler.UpdatedFileInfo"?

jonaskaltenbachz commented:
Hi @Acconut 
any update from your side on this topic?
Should I implement the renaming to "handler.UpdatedFileInfo"?

jonaskaltenbachz commented:
Any update on this topic?

Acconut commented:
@jonaskaltenbachz Thank you very for your persistence here! In the meantime, I implemented some minor refactoring to the handling of response hooks (https://github.com/tus/tusd/commit/d5eca08944f30441618a2b5a0a184ad698222ec4 and https://github.com/tus/tusd/commit/b10c1876b1573d007228eea36f7e8af7ec8bfb86). This made it simpler to add optional return values from the pre-create hook. Based on your PR and our discussion, I then implement a system very similar to what we agreed on in #962. The overall idea is still the same, but names are slightly different and logic is in more appropriate places now.

Thank you very much for your work and dedication here, Jonas! It would be greatly appreciated if you and @genofire could have a look at the new PR.
